arbitrum/apibackend.go

```diff
@@ -433,40 +434,59 @@ func (a *APIBackend) BlockByNumberOrHash(ctx context.Context, blockNrOrHash rpc.
 	return nil, errors.New("invalid arguments; neither block nor hash specified")
 }
 
-func (a *APIBackend) stateAndHeaderFromHeader(ctx context.Context, header *types.Header, err error) (*state.StateDB, *types.Header, error) {
+func (a *APIBackend) stateAndHeaderFromHeader(ctx context.Context, header *types.Header, err error) (*state.StateDB, *types.Header, ethapi.StateReleaseFunc, error) {
```

Comment:
I see why this solution makes sense and it's probably what I'd go for if we were maintaining geth.. but this seems like a merge hell. Maybe it's required but I'm thinking about ways to avoid it..

should we use a finalizer for the returned statedb instead? (not sure this is a good idea, as it's ugly and trusting geth itself will never set a finalizer for it)

Reply:
let's discuss it on the meeting tommorrow, we could go with the finalizer, but what worries me is that there is no guarantee when it will be called and that could be an issue during congestion of RPC calls:

```
// Note that because finalizers may execute arbitrarily far into the future
// after an object is no longer referenced, the runtime is allowed to perform
// a space-saving optimization that batches objects together in a single
// allocation slot. The finalizer for an unreferenced object in such an
// allocation may never run if it always exists in the same batch as a
// referenced object. Typically, this batching only happens for tiny
// (on the order of 16 bytes or less) and pointer-free objects.
```
(https://github.com/golang/go/blob/9db508d46fe201ecc6184ebd7b73770ed38fbdf9/src/runtime/mfinal.go#L369-L376)

Seems that it should "typically" work, as `StateDB` has pointer fields, imo we have to decide whether it is good enough.